Add guidelines for polymorphism (#433)
Microsoft/api-guidelines: azure/Guidelines.md

<a name="json-document-extensible-enums"></a>
:white_check_mark: **DO** document to customers that new values may appear in the future so that customers write their code today expecting these new values tomorrow.

<a name="json-return-extensible-enum-value"></a>
:heavy_check_mark: **YOU MAY** return a value for an extensible enum that is not one of the values defined for the api-version specified in the request.

<a name="json-accept-extensible-enum-value"></a>
:warning: **YOU SHOULD NOT** accept a value for an extensible enum that is not one of the values defined for the api-version specified in the request.

<a name="json-removing-enum-value-is-breaking"></a>

> :no_entry: **DO NOT** remove values from your enumeration list as this breaks customer code.

Reviewer: Just so we're all clear: A service no longer supporting a value is a breaking change and this is allowed following the breaking change policy. In this case, all bets are off though and so the enum value CAN be removed during a formal breaking change.

Author: Yes, this guideline can be violated by declaring a formal breaking change. But I don't think we want to soften the guidance.

We have allowed some breaking change guidance to seep into this doc. In some ways it is unavoidable given that at least parts of our guidance are motivated by avoiding breaking changes later.

If you can't avoid them, then follow the guideline below.

<a name="json-use-discriminator-for-polymorphism"></a>

> :white_check_mark: **DO** define a discriminator field indicating the kind of the resource and include any kind-specific fields in the body.

Reviewer 2: nit, it is a bit hard for me to understand phrase `kind-specific fields`. Would it be better if we use `"kind"-specific fields` or a bit more detailed like `fields relate to this particular "kind" of the resource`?

Reviewer 3: What do we think of:

```suggestion
:white_check_mark: **DO** define a discriminator field indicating the resource kind in the body, alongside fields that are specific to that kind of resource.
```

Reviewer: I'm OK with this wording as it is close enough and most resources have only 1 polymorphic type in them But, in reality, a resource could have multiple polymorphic types in it, for example:
{
   "shapeKind" : "circle",
   "circle" : { ... circle-specific properties... },
   
   "subscriptionKind": "CreditCard",
   "creditCard" : { ... creditCard-specific properties ... },
}

We just need to be aware of this should it come up. I did build a service once many years ago where this was a real issue.

Author: I don't believe it is possible to describe a schema with multiple polymorphic types in OpenAPI. A schema can have at most one `discriminator`.

```
Both Rectangle and Circle have common fields: `kind`, `fillColor`, `lineColor`, and `subscription`. A Rectangle also has `x`, `y`, `width`, and `length` while a Circle has `x`, `y`, and `radius`. The `subscription` is a nested polymorphic type. A `free` subscription has no additional fields and a `paid` subscription has `expiration` and `invoice` fields.

<a name="json-polymorphism-kind-extensible"></a>

> :ballot_box_with_check: **YOU SHOULD** define the kind field of a polymorphic type to be an extensible enum.

Reviewer: We also recommend the field be called 'kind' unless there is a great reason not to - perhaps a guideline on this (maybe in the naming section).

Author: I think that is covered in our [naming guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/ConsiderationsForServiceDesign.md#common-names).

Reviewer 4: Does it hurt to call out here plain as day? We have a lot of guidelines across 2 or 3 files.

Reviewer 4: To clarify, I don't think we should be so explicit everywhere and duplicate information; however, a brief mention of using "kind" is recommended seems trivial in this case.

<a name="json-polymorphism-kind-immutable"></a>

> :warning: **YOU SHOULD NOT** allow an update (patch) to change the kind field of a polymorphic type.

Reviewer: Wait, I'm not sure about this. A client might want to change a circule to a rectangle or vice versa. I think this _has to be_ allowed. So PATCH CAN change the kind field's value as long as it also specifies any properties required by the new kind value.

Author: Are there actual cases in Azure where this is needed?

Reviewer: @johanste would know better than I. 
But, before I joined MS, I worked on a subscription web service customers could change their subscription type from Free to Invoice to Monthly, etc. as there were all polymorphic types. Free had an expiration time period property; Invoice had invoice # and contact email; Subscription had credit card info. So, while I can't recall any specific Azure example, my service would have required this ability.

Reviewer 5: The question is what you do for fields that were previously set on the resource, but that are not valid on the new kind. This is a sometimes a reason to use PUT rather than PATCH.

The scenarios where this has come up is when changing a data source type etc. (where data source types are pluggable)

<a name="json-polymorphism-versioning"></a>
:warning: **YOU SHOULD NOT** return properties of a polymorphic type that are not defined for the api-version specified in the request.

<a name="json-polymorphism-arrays"></a>

> :warning: **YOU SHOULD NOT** specify a polymorphic type as the type of an array property.

Reviewer: I'm not sure what this means. An array of polymorphic types is OK? Or just problematic?
If problematic, then why? 
I do remember we said something about this, but I can't remember the full context.

Author: This is a special case of the known problem that arrays are not "merge-patch-friendly" -- to add or remove an element of an array with merge-patch you must specify the entire array contents, modified accordingly. This is problematic for polymorphic types because it acts like a "replace" on individual elements, so properties of the element that were not returned due to older api-version get lost.

Reviewer: Oh yea, thanks for the refresher. Maybe rephrase this bullet:
**YOU SHOULD NOT** have a property whose value is an array of polymorphic types.

Reviewer 5: So what should you do if you return a sequence of values? For example, Form Recognizer returns a set of recognized fields. They are of different type.

The json-merge-patch gotcha only applies to updatable resources.

Reviewer: Good point. This rule should only apply to updatable resources.

> :warning: **YOU SHOULD NOT** specify a polymorphic type as the type of an array property.

Author: ```suggestion
:warning: **YOU SHOULD NOT** have a property whose value is an array of polymorphic objects.
```

> :warning: **YOU SHOULD NOT** have a property whose value is an array of polymorphic objects.

Reviewer 4: Should the "considerations" doc suggest a pattern so we don't have numerous different ways of doing it across services?
